### PR TITLE
Correctly handle spaces in file/directory names

### LINKED ISSUE #778: tmt run does not handle white space in the path
Having source in a directory with white spaces causes rsync and "tmt run" to fail.

Reproducer:
$ mkdir foo\ bar
$ cd foo\ bar/
$ tmt init
Tree '/tmp/foo bar' initialized.
To populate it with example content, use --template with mini, base or full.
$ tmt run
/var/tmp/tmt/run-001

stderr (3/3 lines)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rsync: link_stat "/tmp/foo" failed: No such file or directory (2)
rsync: change_dir "/var/tmp/tmt/run-003/plans/default//bar" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]

Failed to run command 'rsync -ar --exclude .git /tmp/foo bar/ /var/tmp/tmt/run-003/plans/default/tree'.

### PR DESCRIPTION
**Pull request by psss**
Adjust workdir syncing code and ssh command line generation to
work fine with file paths which contain spaces. The beakerlib
tests still can be broken until the upstream issue is fixed as
well: https://github.com/beakerlib/beakerlib/pull/101

Resolves #778.

### CODE REVIEW

**Comment by lukaszachy** [comment]
Hm, can't we just change run commands to list instead of string (which is later split into cmd + args?)
Because while this fixes space issue, quotes will still break execute. E.g. test called ```/a'a/test``` makes sync fail with `unexpected EOF while looking for matching`

**Comment by psss** [comment]
> can't we just change run commands to list instead of string

Well, could be possible but I'm not able to convince `rsync` to accept the `-e` argument in this way. Do you see anything wrong here?

```python
self.run(
    ["rsync"] + options + ["-e", self._ssh_command(join=True),
    source, f'{self._ssh_guest()}:{destination}'])
```

Yet, the `rsync` complains about invalid syntax.

    rsync error: syntax or usage error (code 1) at main.c(1732) [client=3.2.3]

The command list looks like this:

```
['rsync', '-Rrz', '--links', '--safe-links', '--delete', '-e', 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -p 10048 -i "/var/tmp/tmt/run-110/a directory/some plans/a plan/provision/default/id_rsa"', '/var/tmp/tmt/run-110/a directory/some plans/a plan', 'root@127.0.0.1:/']
```

**Comment by lukaszachy** [comment]
Replied on IRC but in the case someone bumps into this issue - `shell=True` was used

File: tmt/steps/provision/__init__.py [hunk 1]
```diff
@@ -222,14 +222,17 @@ def _ssh_options(self, join=False):
         if self.port:
             options.extend(['-p', str(self.port)])
         if self.key:
-            options.extend(['-i', self.key])
+            options.extend(['-i', f'"{self.key}"' if join else self.key])
         return ' '.join(options) if join else options
```
Comment (lukaszachy) [inline tmt/steps/provision/__init__.py:226]: IMHO it is safer to run shlex.quote on whole options. self.key can contain double quote

File: tmt/steps/provision/__init__.py [hunk 2]
```diff
@@ -222,14 +222,17 @@ def _ssh_options(self, join=False):
...
 
     def _ssh_command(self, join=False):
         """ Prepare an ssh command line for execution (list or joined) """
         command = ['sshpass', f'-p{self.password}'] if self.password else []
-        command += ['ssh'] + self._ssh_options()
-        return ' '.join(command) if join else command
+        command.append("ssh")
+        if join:
+            return " ".join(command) + " " + self._ssh_options(join=True)
```
Comment (lukaszachy) [inline tmt/steps/provision/__init__.py:233]: need to quote - self.password can contains single/double quotes

**Review by lukaszachy — CHANGES_REQUESTED** [review]
Few more places with potential for single/double quotes in the input

File: tmt/steps/provision/__init__.py (hunk 1 above)
Comment (psss) [reply tmt/steps/provision/__init__.py:228]: Good point, substituted with `shlex.quote`.

File: tmt/steps/provision/__init__.py (hunk 2 above)
Comment (psss) [reply tmt/steps/provision/__init__.py:238]: Fixed using `shlex.quote` as well.

**Comment by psss** [comment]
Fedora 33 failure is `Could not resolve host...` issue, known & unrelated (fyi @thrix).

**Review by lukaszachy — APPROVED** [review]

**Pull request merged**